zerver/worker/queue_processors.py

```diff
@@ -101,6 +105,22 @@ def check_and_send_restart_signal():
     except Exception:
         pass
 
+def retry_email(func):
+    # type: (Callable[[QueueProcessingWorker, Mapping[str, Any]], None]) -> Callable[[QueueProcessingWorker, Mapping[str, Any]], None]
+    @wraps(func)
+    def wrapper(worker, data):
+        # type: (QueueProcessingWorker, Mapping[str, Any]) -> None
+        try:
+            func(worker, data)
+        except (smtplib.SMTPServerDisconnected, socket.gaierror):
+            def on_failure(event):
+                # type: (Mapping[str, Any]) -> None
+                logging.exception("Event {} failed".format(event['id']))
+
+            retry_event(worker.queue_name, data, on_failure)
+
+    return wrapper
```

Comment:
This looks reasonable; it'd be great to have a test case that mocks `send_email` to throw one of these exceptions so that we can verify this.  

For testing, I opened #6542 with some notes on where we should be going here.  But probably we can just unit test the `retry_email` decorator and that'll be good enough.  Also, I'd rename it to `retry_send_email_failures`, for clarity.

Reply:
OK.